src/libraries/System.Runtime/tests/System.IO.FileSystem.Tests/RandomAccess/NonSeekable.cs

```diff
@@ -0,0 +1,429 @@
+                Task readToNonEmpty = ReadExactlyAsync(readHandle, buffer, content.Length); // what is required for the above write to succeed
+
+                await Task.WhenAll(readToNonEmpty, write);
+
+                Assert.Equal(content, buffer.AsSpan(0, content.Length).ToArray());
+            }
+        }
+
+        [Fact]
+        public async Task CanReadToStackAllocatedMemory()
+        {
+            (SafeFileHandle readHandle, SafeFileHandle writeHandle) = await GetNamedPipeHandlesAsync();
+
+            using (readHandle)
+            using (writeHandle)
+            {
+                byte[] content = RandomNumberGenerator.GetBytes(BufferSize);
+                Task write = RandomAccess.WriteAsync(writeHandle, content, fileOffset: 0).AsTask();
+
+                ReadToStackAllocatedBuffer(readHandle, content);
+
+                await write;
+            }
+
+            void ReadToStackAllocatedBuffer(SafeFileHandle handle, byte[] array)
+            {
+                Span<byte> buffer = stackalloc byte[array.Length * 2];
+                ReadExactly(handle, buffer, array.Length);
+                Assert.Equal(array, buffer.Slice(0, array.Length).ToArray());
+            }
+        }
+
+        [Fact]
+        public async Task CanWriteFromStackAllocatedMemory()
+        {
+            (SafeFileHandle readHandle, SafeFileHandle writeHandle) = await GetNamedPipeHandlesAsync();
+
+            using (readHandle)
+            using (writeHandle)
+            {
+                byte[] content = RandomNumberGenerator.GetBytes(BufferSize);
+                byte[] buffer = new byte[content.Length * 2];
+                Task read = ReadExactlyAsync(readHandle, buffer, content.Length);
+
+                WriteFromStackAllocatedBuffer(writeHandle, content);
+
+                await read;
+                Assert.Equal(content, buffer.AsSpan(0, content.Length).ToArray());
+            }
+
+            void WriteFromStackAllocatedBuffer(SafeFileHandle handle, byte[] array)
+            {
+                Span<byte> buffer = stackalloc byte[array.Length];
+                array.CopyTo(buffer);
+                RandomAccess.Write(handle, buffer, fileOffset: 0);
+            }
+        }
+
+        [Fact]
+        public async Task FileOffsetsAreIgnored_AsyncWrite_SyncRead()
+        {
+            (SafeFileHandle readHandle, SafeFileHandle writeHandle) = await GetNamedPipeHandlesAsync();
+
+            using (readHandle)
+            using (writeHandle)
+            {
+                byte[] content = RandomNumberGenerator.GetBytes(BufferSize);
+                Task writeToOffset123 = RandomAccess.WriteAsync(writeHandle, content, fileOffset: 123).AsTask();
+                byte[] buffer = new byte[content.Length * 2];
+                int readFromOffset456 = RandomAccess.Read(readHandle, buffer, fileOffset: 456);
+
+                Assert.InRange(readFromOffset456, 1, content.Length);
```

Comment:
Is the InRange used because it is not right to assume that Read will fill the whole buffer?